source/common/config/watch_map.cc

```diff
@@ -93,6 +93,32 @@ void WatchMap::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>
   }
 }
 
+// For responses to on-demand requests, replace the original watch for an alias
```

Comment:
This seems quite complicated. Could watches just monitor a set which looks like {name}.union(aliases)? This is closer to what @markdroth had mentioned. @fredlas WDYT?

Reply:
I mentioned this to @fredlas already: I considered creating watches for aliases and the resource name, An issue with this is that I'd need to keep alias watches in sync with resource aliases (which can change), and would need to do that on every update involving the resource. Since this is a protobuf, the I'd need to iterate over aliases to find matches, which will get expensive quickly.

Considering there's no practical benefit to keep alias watches around (at least at the moment), the simplest thing turned out to be discarding them on an update.

Reply:
> My main concern here is we've introduced complexity by needing to the alias --> named watch conversion.

Yes, but I think you are discarding the complexity of keeping alias watches around. They would need to be kept in sync with virtual hosts: on every update, for every vhost in the update, all its aliases(/domains?) will need to be synchronised. On watch removals, all aliases will need to be removed as well, a way to relate aliases and resource names would need to added and its state maintained.

> ... a model in which names and aliases can be used interchangeably
I don't think this is desirable, at least not with VHDS: aliases there are based on a virtual host data, which is inherently non-constant -- using them as ids would require maintaining and propagating id changes. Aliases were originally created to handle on-demand request/response cycles, perhaps they should be relegated to just that?

Reply:
I think that, at the xDS protocol level, resource names and aliases should be fully interchangeable.  Specifically, any string used to identify a resource (regardless of whether it is present in the name field or in the aliases list) must be unique for that resource type, and any subscription request for a given string should match either a resource whose name is the specified string or a resource that has an alias matching the specified string.

I'm not at all familiar with the Envoy code here, but I would think that this could be accomplished by storing the data as something like `std::map<std::string /* name or alias */, std::shared_ptr<Resource>>`.  When we get a new resource from the server, we add an entry to that map for the name and for every alias, all pointing to the same `Resource` object.  If the resource had already previously existed, then we identify any entries in the map that are no longer relevant and remove them.

Reply:
I think at this point, in terms of internal implementation, I'd like to have those folks who have made the investment in the architecture and ongoing maintenance to make the call on whether we do alias watches or the conversion at this point. Either works, I'd like to ensure we have a sustainable and clean approach going forward. @fredlas @lambdai what are your thoughts?

Reply:
@fredlas shared offline that he thinks this approach is sound.

Reply:
I don't have a strong opinion on this one.

Reply:
Dmitri convinced me pretty well in https://github.com/envoyproxy/envoy/pull/8617#discussion_r338838148 that this will all work cleanly enough. Also, zooming out a bit, I think conversion, as opposed to building understanding of aliases into the watch machinery, is the right general direction. The whole watch/subscription setup should try to avoid more complexity; it feels like it's currently doing just enough while staying mostly comprehensible. With conversion, the users of this exotic alias stuff have a single point in the code where they get translated back to "normal". Although it's going to live in watch_map.cc, it's an entirely self-contained function, so easy enough to ignore.